### PR TITLE
fix: OnRequest should wait all readable data consumed when sender close connection

### PR DESCRIPTION
**Pull request by joway**
#### What type of PR is this?

fix

#### Check the PR title.

- [x] This PR title match the format: \<type\>(optional scope): \<description\>
- [x] The description of this PR title is user-oriented and clear enough for others to understand.
- [x] Attach the PR updating the user documentation if the current PR requires user awareness at the usage level. [User docs repo](https://github.com/cloudwego/cloudwego.github.io)

#### (Optional) More detailed description for this PR(en: English/zh: Chinese).

en: Fix two issues:
1. OnRequest should wait all readable data consumed when peer close connection.
2. When peer close connection, the error should be EOF not ErrConnClosed.
3. fix "detach no such file or directory" error
4. fix operator race
zh(optional):

### CODE REVIEW

File: connection_reactor.go [hunk 1]
```diff
@@ -25,32 +25,44 @@ import (
...
 		c.closeCallback(true)
 		return nil
 	}
-	if c.isCloseBy(poller) {
-		// Connection with OnRequest of nil
-		// relies on the user to actively close the connection to recycle resources.
+
+	closedByPoller := c.isCloseBy(poller)
+	// force change closed by user
+	c.force(closing, user)
```
Comment (jayantxie) [inline connection_reactor.go:62]: 这个是解决什么问题？

File: net_polldesc.go [hunk 2]
```diff
@@ -21,16 +21,15 @@ import (
...
 func newPollDesc(fd int) *pollDesc {
 	pd := &pollDesc{}
 	poll := pollmanager.Pick()
-	op := poll.Alloc()
-	op.FD = fd
-	op.OnWrite = pd.onwrite
-	op.OnHup = pd.onhup
-
-	pd.operator = op
+	pd.operator = &FDOperator{
```
Comment (jayantxie) [inline net_polldesc.go:27]: 直接走堆分配，如果关闭连接解除引用，poller内部会有问题吧？

File: net_polldesc.go [hunk 3]
```diff
@@ -84,6 +76,7 @@ func (pd *pollDesc) onwrite(p Poll) error {
 	select {
 	case <-pd.writeTrigger:
 	default:
+		pd.detach()
```
Comment (jayantxie) [inline net_polldesc.go:79]: 这是为啥要detach掉？

File: poll_default_bsd.go [hunk 4]
```diff
@@ -105,21 +106,28 @@ func (p *defaultPoll) Wait() error {
...
+					var leftRead int
+					// read all left data if peer send and close
+					if leftRead, err = readall(operator, barriers[i]); err != nil && !errors.Is(err, ErrEOF) {
+						logger.Printf("NETPOLL: readall(fd=%d)=%d before close: %s", operator.FD, total, err.Error())
+					}
+					totalRead += leftRead
+				}
+				// only close connection if no further read bytes
+				if totalRead == 0 {
+					p.appendHup(operator)
```
Comment (jayantxie) [inline poll_default_bsd.go:128]: 如果不调用AppendHup，poller就无法detach连接，这个是会导致poller反复触发的吧？

Comment (joway) [reply poll_default_bsd.go:128]: linux 下这个是 ET 模式，不会重复触发。此外，哪怕重复，下一次poller就没数据可以读了，就会直接关闭了

File: net_polldesc.go (hunk 2 above)
Comment (joway) [reply net_polldesc.go:27]: 这个只有 client 创建连接时候出现，而用户要关闭这个连接，必须等到client创建完连接，也就是说这个operator已经彻底用完了。所以不会有问题

File: connection_reactor.go (hunk 1 above)
Comment (joway) [reply connection_reactor.go:62]: 遇到对端关闭连接时，有两个阶段：
1. poller 拿到 poll 事件，把连接关闭了，但这个阶段是软性关闭，不释放资源，因为连接可以依然被其他东西正在操作
2. user 遇到比如 read 错误，知道了连接已经关闭了，这个时候会主动调用 conn.Close() 。标识我知道了，且保证之后不再用这个连接了，此时我们才能够释放资源。

File: net_polldesc.go (hunk 2 above)
Comment (jayantxie) [reply net_polldesc.go:27]: OK，这个没问题

File: net_polldesc.go (hunk 3 above)
Comment (jayantxie) [reply net_polldesc.go:79]: 如果这一步不detach，直接重新注册，是不是也是ok的?

File: poll_default_bsd.go (hunk 4 above)
Comment (jayantxie) [reply poll_default_bsd.go:128]: get

File: connection_reactor.go (hunk 1 above)
Comment (jayantxie) [reply connection_reactor.go:62]: get

File: net_polldesc.go (hunk 3 above)
Comment (joway) [reply net_polldesc.go:79]: @jayantxie 如果能够100%对operator做好保护的话，那就可以。但是目前看，这个复杂性太高了，而且每一次保护都是一个院子操作，得不偿失

**Review by jayantxie — APPROVED** [review]
